design/one-pager-template-stacks.md

```diff
@@ -0,0 +1,679 @@
+are immutable, so there is no assurance concern here.
+
+What of the CRDs? Revised CRDs in an updated Stack will be updated because they
+are unique, but how will Stack CRDs that have been removed be deleted? This is a
+trickier problem to be considered in future designs.
+
+The TSM should set a Controller Reference on all template rendered resources.
+The appropriate Controller Reference in this case would be the CR of the managed
+type. For example, if a Stack defined a `Wordpress` type whose template included
+a `MySQLInstance`, that `MySQLInstance` would use `Wordpress` as the Controller
+Reference. If `Wordpress` is deleted, the `MySQLInstance` will be deleted (based
+on reclaim policy).
+
+The TSM should make the `Stack` the [Owner
+Reference](https://kubernetes.io/docs/concepts/workloads/controllers/garbage-collection/#owners-and-dependents)
+for any managed instances of the managed kind (`Wordpress`). This will ensure
+that when the Stack is removed, all managed resources created from that Stack
+are also removed.
+
+### RBAC
+
+The Stack Manager currently creates a Service Account for Stack Deployments,
+restricted by the Owned and Dependent types declared by the Stack. A Template
+Stack Manager will benefit from this same processing with little or no changes
+needed. The roles for each TSM controller will be tailored to the types that
+Stack needs to handle and render. The TSM will also need a `Role` permitting it
+to read the `Stack` it is operating on. It is important for each TS to take
+advantage of the Stack `spec.dependsOn` field to define all of the resources
+defined in the template.
+
+An alternative to deploying a TSM per Stack would be to have the existing Stack
+Manager handle all new TSM defined types. The SM runs with cluster privileges,
+giving Template Stacks to much control. Under this approach, the existing Stack
+Manager could still create Service Accounts and take advantage of User
+Impersonation when reconciling for types defined within each Template Stack.
+This approach would not offer simple visibility into the roles the deployment is
+actively using.
+
+By creating a TSM deployment for each Template Stack that is installed, we are
+able to scope the permissions given to each TSM instance to the permissions
+requested by the Stack (i.e. the CRDs it owns and depends on). This is analogous
+to how the Stack Manager currently creates a `Deployment` to host each
+controller-runtime based Stack’s controller which is also scoped down to the
+needed permissions.
+
+#### Concerning the use of Core Types
+
+In several examples throughout this design, Template Stacks are demonstrated
+generating `Deployment` resources by including them in template text. There are
+compelling justifications for both permitting and denying Template Stacks from
+declaring a dependency on core and extension types, such as
+`deployment.apps/v1`.
+
+The ability to use Template Stacks to render arbitrary Kubernetes resources is a
```

Comment:
I won't resolve this thread until there is an issue to track the use of core types.

I think we are going to have a much easier time creating template stacks if we don't need to create a KubernetesApplication for each independent resource.

Take 
https://gitlab.com/gitlab-org/charts/gitlab/blob/master/charts/gitlab/charts/operator/templates/operator_deployment.yaml
and the neighboring templates for example.  We could make a set of Template Stacks, each representing one of the Gitlab CRDs, or a single Template Stack representing all of the CRDs (depending on how much control (roles) you want to give the TSM(s)).

Either way, there will be lots of core resources needed.  Does each Gitlab Template Stack put all of these core resources in a single KubernetesApplication resource within the Stack?

Does each Template Stack resource (or template) then require a KubernetesCluster reference?

I think we'll need a real world example to determine if Crossplane managed types *CAN* be sufficient to translate existing chart projects to Template Stacks.

I'm reluctant to assert that Template Stacks can only handle Crossplane managed types until we can prove that this offers sufficient capability in practice.

I think this section of the design already exposes both ideologies, deferring to future design work.  To your point, I could clarify by concluding the section stating core types are not expressly blocked by this design and the current Stacks implementation, but this may not always be the case.

Reply:
Looking at those Gitlab charts has made me wonder if we may also need a third field:
* `templates` K8s resource template text keyed by both GVK, then a unique template name
* `templateStatus` sub-resource template text keyed by GVK
* `includes` (not really this name...) arbitrary templates that are not necessarily k8s resources or sub-resources, which could be reused by `templates` or `templateStatus`.
   I don't know if these should be shared between CRDs (how could they if they include CRD specific template variables).
   These could perhaps live under `templates`, just using arbitrary paths / names instead of CRD names.